app/code/Magento/CatalogUrlRewrite/Observer/ProductProcessUrlRewriteSavingObserver.php

```diff
@@ -65,11 +117,46 @@ public function execute(\Magento\Framework\Event\Observer $observer)
             || $product->getIsChangedWebsites()
             || $product->dataHasChangedFor('visibility')
         ) {
-            if ($product->isVisibleInSiteVisibility()) {
-                $product->unsUrlPath();
-                $product->setUrlPath($this->productUrlPathGenerator->getUrlPath($product));
+            //Refresh rewrite urls
+            $product->unsUrlPath();
+            $product->setUrlPath($this->productUrlPathGenerator->getUrlPath($product));
+            if (!empty($this->productUrlRewriteGenerator->generate($product))) {
                 $this->urlPersist->replace($this->productUrlRewriteGenerator->generate($product));
             }
+
+            $storeIdsToRemove = [];
+            if ($this->productScopeRewriteGenerator->isGlobalScope($product->getStoreId())) {
+                //Remove any rewrite URLs for websites the product is not in, or is not visible in. Global Scope.
+                foreach ($this->storeManager->getWebsites() as $website) {
+                    $websiteId = $website->getWebsiteId();
+                    foreach ($this->storeWebsiteRelation->getStoreByWebsiteId($websiteId) as $storeid) {
+                        //Load the product for the store we are processing so we can see if it is visible
+                        $storeProduct = $this->productRepository->getById(
```

Comment:
This looks as really heavy operation to load whole product object with all attributes, extension attributes, etc in this case. Can we get this info by doing some query?

I didn't ran any performance tests, but I think it will significantly slow down product save on multi store websites

Reply:
@ihor-sviziev What do you think to something like this

```
$storeProduct = $productCollection->addAttributeToFilter('entity_id', $product->getId())
                        ->addAttributeToSelect(['visibility','entity_id'])
                        ->setPage(1,1)
                        ->setStoreId($storeId)
                        ->getFirstItem();
```
That gives a slimmed down product

![image](https://user-images.githubusercontent.com/30697781/76014315-4da58f80-5f11-11ea-8e32-0460d91b6170.png)

I couldn't think of a better way of doing it?

Reply:
I think for this specific case this solution would be much much better

Reply:
OK, perhaps it isn't related to database transactions. I've asked on Slack for help  https://magentocommeng.slack.com/archives/C4YS78WE6/p1583489502082300

Reply:
Right, I found a much more pleasing way to get the visibilities for all products in all websites in a single call.

`$storeVisibilities = $this->productCollection->getAllAttributeValues(ProductInterface::VISIBILITY);`

This returns an array of ['productId']['storeId']['visibility']

Commit coming shortly